### PR TITLE
Add Dual-Stack support to L4 NetLB

### PR DESCRIPTION
**Pull request by panslava**
Add support for `spec.ipFamilies` and `spec.ipFamiliyPolicy` in L4 NetLB

- Add new feature-flag `--enable-l4netlb-dual-stack`, all the new logic is hidden behind it
- Support all the transitions between ipFamilies, controller sync will provide resources for required ipFamily and clean resources for not required, but only if they exist in annotations
- On service deletion, clean up all the resources, no matter if they exist in annotations

### CODE REVIEW

**Comment by cezarygerard** [comment]
Please ty to split this PR into few smalle ones

File: pkg/loadbalancers/forwarding_rules_ipv6.go [hunk 1]
```diff
@@ -127,6 +127,79 @@ func (l4 *L4) deleteChangedIPv6ForwardingRule(existingFwdRule *composite.Forward
 	return nil
 }
 
+func (l4netlb *L4NetLB) ensureIPv6ForwardingRule(bsLink string) (*composite.ForwardingRule, error) {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/forwarding_rules_ipv6.go:131]: done

File: pkg/l4lb/l4netlbcontroller_test.go [hunk 2]
```diff
@@ -1328,3 +1407,74 @@ func TestIsRBSBasedServiceForNonLoadBalancersType(t *testing.T) {
...
+			svc, err = controller.ctx.KubeClient.CoreV1().Services(svc.Namespace).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("Failed to lookup service %s, err %v", svc.Name, err)
+			}
+
+			expectedIngressLength := len(tc.ipFamilies)
+			// For empty IP Families we should provide ipv4 address
+			if expectedIngressLength == 0 {
+				expectedIngressLength = 1
+			}
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/l4lb/l4netlbcontroller_test.go:1468]: decided offline to leave it

File: pkg/loadbalancers/l4netlb.go [hunk 3]
```diff
@@ -180,53 +284,122 @@ func (l4netlb *L4NetLB) EnsureLoadBalancerDeleted(svc *corev1.Service) *L4NetLBS
...
-		klog.V(2).Infof("Finished deleting nodes firewall %s for L4 NetLB Service %s/%s, time taken: %v", firewallName, l4netlb.Service.Namespace, l4netlb.Service.Name, time.Since(start))
-	}()
+// deleteIPv4ResourcesOnDelete deletes all resources specific to IPv4.
+// This function is called only on Service deletion.
+// During sync, we delete resources only that exist in annotations,
+// so they could be leaked, if annotation was deleted.
+// That's why on service deletion we delete all IPv4 resources, ignoring their existence in annotations
+func (l4netlb *L4NetLB) deleteIPv4ResourcesOnDelete(result *L4NetLBSyncResult) {
+	klog.Infof("Deleting IPv4 resources for L4 NetLB Service %s/%s on delete, without checking for existence in annotation", l4netlb.Service.Namespace, l4netlb.Service.Name)
+	l4netlb.deleteIPv4ResourcesAnnotationBased(result, true)
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb.go:316]: We decided offline to leave it as it is

File: pkg/loadbalancers/forwarding_rules_ipv6.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/forwarding_rules_ipv6.go:131]: it is tested in l4netlb_test.go

File: pkg/loadbalancers/l4netlb_test.go [hunk 4]
```diff
@@ -217,7 +198,407 @@ func TestHealthCheckFirewallDeletionWithILB(t *testing.T) {
 	if err == nil || !strings.Contains(err.Error(), "not found") {
 		t.Errorf("Healthcheck %s should be deleted", hcName)
 	}
+}
+
+func TestMetricsForStandardNetworkTier(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go:203]: My pr only moved this test from the bottom of the file

Discussed offline to leave it

File: pkg/loadbalancers/l4netlbipv6.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		return
+	}
+
+	if ipv6fr.IPProtocol == string(corev1.ProtocolTCP) {
+		syncResult.Annotations[annotations.TCPForwardingRuleIPv6Key] = ipv6fr.Name
+	} else {
+		syncResult.Annotations[annotations.UDPForwardingRuleIPv6Key] = ipv6fr.Name
+	}
+
+	l4netlb.ensureIPv6NodesFirewall(ipv6fr, nodeNames, syncResult)
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlbipv6.go]: done

File: pkg/test/utils.go [hunk 6]
```diff
@@ -153,6 +153,18 @@ func NewL4NetLBRBSService(port int) *api_v1.Service {
 	return svc
 }
 
+// NewL4NetLBRBSDualStackService creates a Service of type LoadBalancer with RBS Annotation and provided ipFamilies and ipFamilyPolicy
+func NewL4NetLBRBSDualStackService(port int, protocol api_v1.Protocol, ipFamilies []api_v1.IPFamily, externalTrafficPolicy api_v1.ServiceExternalTrafficPolicyType) *api_v1.Service {
```
Comment (cezarygerard) [inline pkg/test/utils.go:157]: do you really need a port here?

File: pkg/l4lb/l4lbcommon.go [hunk 7]
```diff
@@ -95,6 +95,16 @@ func deleteL4RBSAnnotations(ctx *context.ControllerContext, svc *v1.Service) err
 	return patch.PatchServiceObjectMetadata(ctx.KubeClient.CoreV1(), svc, *newObjectMeta)
 }
 
+// deleteL4RBSDualStackAnnotations deletes all annotations which could be added by L4 ELB RBS controller
+func deleteL4RBSDualStackAnnotations(ctx *context.ControllerContext, svc *v1.Service) error {
+	newObjectMeta := computeNewAnnotationsIfNeeded(svc, nil, loadbalancers.L4DualStackResourceRBSAnnotationKeys)
+	if newObjectMeta == nil {
+		return nil
+	}
+	klog.V(3).Infof("Deleting all DualStack annotations for L4 ELB RBS service %v/%v", svc.Namespace, svc.Name)
```
Comment (cezarygerard) [inline pkg/l4lb/l4lbcommon.go:104]: what verbosity we use in GCP? 

Let's make sure this is logged

File: pkg/loadbalancers/l4netlb.go [hunk 8]
```diff
@@ -124,54 +126,156 @@ func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service)
 
 	l4netlb.Service = svc
 
-	sharedHC := !helpers.RequestsOnlyLocalTraffic(svc)
+	hcLink := l4netlb.provideHealthChecks(nodeNames, result)
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4netlb.go:129]: I feel like we have very few debug logs from EnsureFrontend and from the functions called from EnsureFrontend

**Comment by cezarygerard** [comment]
LGTM overall

Great job @panslava !

File: pkg/loadbalancers/l4_test.go [hunk 9]
```diff
@@ -1726,73 +1723,15 @@ func TestDualStackLoadBalancerTransitions(t *testing.T) {
...
-								l4.Service = finalSvc
-
-								result = l4.EnsureInternalLoadBalancer(nodeNames, svc)
-								finalSvc.Annotations = result.Annotations
-								assertDualStackILBResources(t, l4, nodeNames)
-
-								l4.EnsureInternalLoadBalancerDeleted(l4.Service)
-								assertDualStackILBResourcesDeleted(t, l4)
-							})
+							runL4ILBDualStackTransitionTest(
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4_test.go:1726]: Thanks, done

File: pkg/loadbalancers/l4_test.go [hunk 10]
```diff
@@ -1801,109 +1740,192 @@ func TestDualStackLoadBalancerTransitions(t *testing.T) {
...
+		toProtocol        string
+	}{
+		strings.Join(stringInitialIPFamily, ","),
+		strings.Join(stringFinalIPFamily, ","),
+		string(initialTrafficPolicy),
+		string(finalTrafficPolicy),
+		string(initialProtocol),
+		string(finalProtocol),
+	}
+	return fmt.Sprintf("%v", desc)
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4_test.go:1813]: Thanks, done

File: pkg/loadbalancers/l4netlb.go (hunk 8 above)
Comment (panslava) [reply pkg/loadbalancers/l4netlb.go:129]: Maybe we could add more logging to upper level functions,

But actually, if you go down to any gcloud resource creation, there are logs now for every resource, with measuring, how much time it took

File: pkg/l4lb/l4lbcommon.go (hunk 7 above)
Comment (panslava) [reply pkg/l4lb/l4lbcommon.go:104]: We use -v 3, so it will be logged

**Pull request merged**